activerecord/lib/active_record/associations/builder/has_one.rb

```diff
@@ -32,27 +32,22 @@ def self.define_validations(model, reflection)
       end
     end
 
-    def self.touch_record(o, name, touch)
-      record = o.send name
+    def self.touch_record(record, name, touch)
+      instance = record.send(name)
 
-      return unless record && record.persisted?
-
-      if touch != true
-        record.touch(touch)
-      else
-        record.touch
+      if instance&.persisted?
+        touch != true ?
+          instance.touch(touch) : instance.touch
       end
     end
 
     def self.add_touch_callbacks(model, reflection)
       name  = reflection.name
       touch = reflection.options[:touch]
 
-      callback = lambda { |record|
-        HasOne.touch_record(record, name, touch)
-      }
-
+      callback = -> (record) { HasOne.touch_record(record, name, touch) }
       model.after_create callback, if: :saved_changes?
+      model.after_create_commit { association = association(name); association.reset if association.loaded? && association.target.nil? }
```

Comment:
@kamipo thanks for the test! I've tried to restate the problem to this. I'm not suggesting we ship this, but I was having a hard time getting has_one's `stale_target?` to do the right thing.

I think `association.reset if association.loaded? && association.target.nil?` is pretty good statement of the heart of the issue. That if we've tried to access the association within the transaction, and found nothing, we're never given another chance on the instance unless we do a wholesale `reload`.

Do you have ideas for fixing `stale_target?`/`stale_state` on has_one? I'm drawing blanks because I can't access the "outside of transaction" state from within there.

Reply:
I have a plan to improve automatic inverse_of for fixing existing issues (#35204, #32826, etc). That may mitigate this issue in this case.

Potentially, detecting loaded association target is staled or not is hard thing for foreign associations (has_one, has_many).

I have no idea for fixing `stale_target?`/`stale_state` for has_one, we have a way to do `employable.reload_chef` for staled has_one target (since it is not only issue for has_one touching here) though, but clearing negative target cache in this context seems good.

Reply:
> I have a plan to improve automatic inverse_of for fixing existing issues (#35204, #32826, etc). That may mitigate this issue in this case.

Awesome 🙏

> Potentially, detecting loaded association target is staled or not is hard thing for foreign associations (has_one, has_many).

Very true! 😄

Perhaps the other side should be able to mark the foreign association stale? E.g. after the create transaction commits.

Reply:
> Perhaps the other side should be able to mark the foreign association stale?

Maybe so, but the negative staled target cache in this case is not happened if inverse_of is worked.

Reply:
Yes, you're right, having automatic inverse of work for polymorphic association would be even better. Currently, it seems like automatic inverse of works at the reflection level, but if it could also work when assigning a record we'd fix this.

E.g. if Comment has commentable but `Comment.new(commentable: Post.new)` should be able to inverse the association and assign `post.comment = comment`.